lib/task_listener.js

```diff
@@ -13,166 +13,254 @@ var request = require('superagent-promise');
+    var cancelListener = new taskcluster.PulseListener({
+      credentials: this.runtime.pulse
     });
 
-    yield listener.bind(queueEvents.taskPending({
+    await cancelListener.bind(queueEvents.taskException({
+      workerId: this.runtime.workerId,
       workerType: this.runtime.workerType,
+      workerGroup: this.runtime.workerGroup,
       provisionerId: this.runtime.provisionerId
     }));
 
-    debug('bind task pending', {
-      workerType: this.runtime.workerType,
-      provisionerId: this.runtime.provisionerId
-    });
+    cancelListener.on('message', this.cancelTask.bind(this));
+    await cancelListener.resume();
+    return cancelListener;
+  }
 
-    debug('listen', {
-      queueName: listener._queueName, capacity: this.runtime.capacity
-    });
+  async getTasks() {
+    // Number of tasks we could claim
+    let availabileCapacity = this.capacity - this.pending;
+    if (availabileCapacity <= 0) { return;}
 
-    var channel = yield listener.connect();
-
-    // Rather then use `.consume` on the listener directly we use the channel
-    // directly for greater control over the flow of messages.
-    yield channel.consume(listener._queueName, co(function* (msg) {
-      self.runtime.log('listener begin consume');
-      var content;
-      try {
-        self.incrementPending();
-        // All content from taskcluster should be a json payload.
-        content = JSON.parse(msg.content);
-        yield self.runTask(content);
-        channel.ack(msg);
-        // Only indicate a completed task (which may trigger an idle state)
-        // after an ack/nack.
-        self.decrementPending();
-      } catch (e) {
-        if (content) {
-          self.runtime.log('task error', {
-            taskId: content.status.taskId,
-            runId: content.runId,
-            message: e.toString(),
-            stack: e.stack,
-            err: e
-          });
-        } else {
-          self.runtime.log('task error', {
-            message: e.toString(),
-            err: e
+    let taskQueue = this.taskQueue;
+    debug(`polling for ${availabileCapacity} tasks`);
+
+    let claims = [];
+    let queues = await taskQueue.getQueues();
+
+    for(let queue of queues) {
+      if (claims >= availabileCapacity) break;
+      // Keep polling queue until enough tasks were claimed or there are no more
+      // tasks in the queue
+      // Move onto the next queue if more tasks are needed but current queue is exhausted
+      // This ensures that as many tasks as possible are consumed from the highest priority
+      // queue
+      while(claims.length < availabileCapacity) {
+        let tasksNeeded = availabileCapacity - claims.length;
+        let tasks = await taskQueue.getTasksFromQueue(queue, availabileCapacity-claims.length);
+        if (!tasks.length) break;
+        let newClaims = await Promise.all(tasks.map(async (task) => {
+          return await this.claimTask(task);
+        }));
```

Comment:
Not all `queue.claimTask` are successful... Some returns a 4xx error that should be ignored. Please do filter out null properties from the `newClaims` array... Refer to my comment on `this.claimTask` about explicitely returning null.